backends/dpdk/dpdkAsmOpt.cpp

```diff
@@ -163,4 +163,32 @@ const IR::IndexedVector<IR::DpdkAsmStatement> *RemoveLabelAfterLabel::removeLabe
     }
     return new_l;
 }
+
+const IR::Node* RemoveUnusedMetadataFields::preorder(IR::DpdkAsmProgram *p) {
+    auto usedFieldsCollector = new CollectUsedMetadataField();
+    usedFieldsCollector->setCalledBy(this);
+    p->apply(*usedFieldsCollector);
+    auto used_fields = usedFieldsCollector->getUsedFields();
+
+    IR::IndexedVector<IR::DpdkStructType> usedStruct;
+    for (auto st : p->structType) {
+        IR::IndexedVector<IR::StructField> usedMetadataFields;
+        for (auto field : st->fields) {
+            if (used_fields.count("m." + field->name.name)) {
+                usedMetadataFields.push_back(field);
+            }
+        }
+        if (!usedMetadataFields.empty()) {
+            auto newSt = new IR::DpdkStructType(st->srcInfo, st->name,
+                                             st->annotations, usedMetadataFields);
+            usedStruct.push_back(newSt);
+        }
+        else {
```

Comment:
If none of the metadata fields are used, do you intend to return the original struct? We should be creating an empty struct in that.

Reply:
done